node/coinstacks/solana/api/src/websocket.ts

```diff
@@ -0,0 +1,236 @@
+  jsonrpc: string
+  id: string
+  method: string
+  params?: unknown
+}
+
+export interface Args {
+  geyserUrl: string
+  apiKey?: string
+  blockHandler: BlockHandler | Array<BlockHandler>
+  transactionHandler: TransactionHandler
+}
+
+export interface Options {
+  pingInterval?: number
+  retryAttempts?: number
+}
+
+export type TransactionHandler = (data: GeyserResultTransaction) => Promise<void>
+export type BlockHandler = (data: number) => Promise<void>
+
+export class SolanaWebsocketClient implements IWebsocketClient {
+  private socket: WebSocket
+  private geyserSocket: WebSocket
+  private url: string
+  private geyserUrl: string
+  private pingTimeout?: NodeJS.Timeout
+  private geyserPingTimeout?: NodeJS.Timeout
+  private interval?: NodeJS.Timeout
+  private geyserInterval?: NodeJS.Timeout
+  private retryCount = 0
+  private geyserRetryCount = 0
+  private addresses: Array<string>
+
+  private handleBlock: BlockHandler | Array<BlockHandler>
+  private handleTransaction: TransactionHandler | Array<TransactionHandler>
+
+  private readonly pingInterval: number
+  private readonly retryAttempts: number
+
+  private logger = new Logger({ namespace: ['unchained', 'solana', 'websocket'], level: process.env.LOG_LEVEL })
+
+  constructor(url: string, args: Args, opts?: Options) {
+    this.addresses = []
+    this.handleBlock = args.blockHandler
+    this.handleTransaction = args.transactionHandler
+    this.url = args.apiKey ? `${url}/?api-key=${args.apiKey}` : url
+    this.geyserUrl = args.apiKey ? `${args.geyserUrl}/?api-key=${args.apiKey}` : args.geyserUrl
+    this.socket = new WebSocket(this.url, { handshakeTimeout: 5000 })
+    this.geyserSocket = new WebSocket(this.geyserUrl, { handshakeTimeout: 5000 })
+
+    this.pingInterval = opts?.pingInterval ?? 10000
+    this.retryAttempts = opts?.retryAttempts ?? MAX_RETRY_ATTEMPTS
+
+    this.initialize()
+  }
+
+  private initialize(): void {
+    this.socket.on('ping', () => this.socket.pong())
+    this.socket.on('pong', () => this.heartbeat())
+    this.socket.onerror = (error) => {
+      this.logger.error({ error, fn: 'ws.onerror' }, 'websocket error')
+    }
+    this.socket.onclose = ({ code, reason }) => {
+      this.logger.error({ code, reason, fn: 'ws.close' }, 'websocket closed')
+      this.close()
+    }
+    this.socket.onopen = () => this.onOpen()
+    this.socket.onmessage = (msg) => this.onMessage(msg)
+
+    this.geyserSocket.on('ping', () => this.geyserSocket.pong())
+    this.geyserSocket.on('pong', () => this.geyserHeartbeat())
+    this.geyserSocket.onerror = (error) => {
+      this.logger.error({ error, fn: 'ws.onerror' }, 'websocket error')
+    }
+    this.geyserSocket.onclose = ({ code, reason }) => {
+      this.logger.error({ code, reason, fn: 'ws.close' }, 'websocket closed')
+      this.closeGeyser()
+    }
+    this.geyserSocket.onopen = () => this.onGeyserOpen()
+    this.geyserSocket.onmessage = (msg) => this.onMessage(msg)
+  }
```

Comment:
A lot of duplicates because of the needs of 2 websockets (geyser and regular)

We might be able to fix this by supporting multiple sockets somehow using a better implementation